flink-core/src/main/java/org/apache/flink/core/fs/SafetyNetCloseableRegistry.java

```diff
@@ -99,42 +120,61 @@ protected void doUnRegister(
 			implements Closeable {
 
 		private final Closeable innerCloseable;
+		private final SafetyNetCloseableRegistry closeableRegistry;
 		private final String debugString;
 
 		public PhantomDelegatingCloseableRef(
 				WrappingProxyCloseable<? extends Closeable> referent,
+				SafetyNetCloseableRegistry closeableRegistry,
 				ReferenceQueue<? super WrappingProxyCloseable<? extends Closeable>> q) {
 
 			super(referent, q);
 			this.innerCloseable = Preconditions.checkNotNull(WrappingProxyUtil.stripProxy(referent));
+			this.closeableRegistry = Preconditions.checkNotNull(closeableRegistry);
 			this.debugString = referent.toString();
 		}
 
-		public Closeable getInnerCloseable() {
-			return innerCloseable;
-		}
-
 		public String getDebugString() {
 			return debugString;
 		}
 
 		@Override
 		public void close() throws IOException {
+			synchronized (closeableRegistry.getSynchronizationLock()) {
+				closeableRegistry.closeableToRef.remove(innerCloseable);
```

Comment:
Removal is done here now, because the reaper thread should not keep/locate right references for all registries